lib/certificate.js

```diff
@@ -0,0 +1,145 @@
+    <!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">
+    <plist version=\"1.0\">
+    <array/>
+</plist>`;
+
+/**
+ * Library for programatically adding certificates 
+ */
+class Certificate {
+
+  constructor (pemFilename) { 
+    this.pemFilename = pemFilename;
+  }
+
+  /**
+   * Add a certificate to the TrustStore
+   */
+  async add (dir) {
+    let data = await this.getDerData(this.pemFilename);
+    let subject = await this.getSubject(this.pemFilename);
+    let fingerprint = await this.getFingerPrint(this.data);
+
+    let trustStore = new TrustStore(dir);
+    return trustStore.addRecord(fingerprint, tset, subject, data);
+  }
+
+  /**
+   * Checks if keychain at given directory has this certificate
+   */
+  async has (dir) {
+    let subject = await this.getSubject(this.pemFilename);
+
+    let trustStore = new TrustStore(dir);
+    let records = await trustStore.getRecords(subject);
+    return records.length > 0;
+  }
+
+  /**
+   * Remove certificate from the TrustStore
+   */
+  async remove (dir) { 
+    let subject = await this.getSubject(this.pemFilename);
+    let trustStore = new TrustStore(dir);
+    return trustStore.removeRecord(subject);
+  }
+
+  /**
+   * Translate PEM file to DER buffer
+   */
+  async getDerData () {
+    if (this.data) {
+      return this.data;
+    }
+
+    // Convert 'pem' file to 'der'
+    this.data = await openssl('x509', { 
+      outform: 'der',
+      in: this.pemFilename
+    });
+
+    return this.data;
+  }
+
+  /**
+   * Get SHA1 fingerprint from der data before
+   */
+  async getFingerPrint () {
+    if (this.fingerprint) {
+      return this.fingerprint;
+    }
+
+    let data = await this.getDerData();
+    let shasum = crypto.createHash('sha1');
+    shasum.update(data);
+    this.fingerprint = shasum.digest();
+    return this.fingerprint;
+  }
+
+  /**
+   * Parse the subject from the der data
+   */
+  async getSubject () {
+    if (this.subject) {
+      return this.subject;
+    }
+      
+    // Convert 'pem' file to 'der'
+    let subject = await openssl('x509', {
+      noout: true,
+      subject: true,
+      in: this.pemFilename
+    });
+    let subRegex = /^subject[\w\W]*\/CN=([\w\W]*)(\n)?/;
+    this.subject = subject.toString().match(subRegex)[1]; 
+    return this.subject;
+  }
+
+}
+
+/**
+ * Interface for adding and removing records to TrustStore.sqlite3 databases that Keychains use
+ */
+class TrustStore {
+  constructor (sharedResourceDir) {
+    this.sqliteDBPath = path.resolve(sharedResourceDir, 'Library/Keychains/TrustStore.sqlite3');
+    this.db = new sqlite3.Database(this.sqliteDBPath);
+  }
+
+  /**
+   * Add record to tsettings
+   */
+  async addRecord (sha1, tset, subj, data) {
+    let existingRecords = await this.getRecords(subj);
+    if (existingRecords.length > 0) {
+      return this.db.runAsync(`UPDATE tsettings SET sha1=?, tset=?, data=? WHERE subj=?`, [sha1, tset, data, subj]);
```

Comment:
I'll add 'await' just to keep it consistent. I did actually have a race condition in previous tests because I wasn't returning it. I could see how there'd be a race condition if the addRecord is called with awaiting the result.